src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,241 @@
+                # means it's a symlink but not to the ctdb
+                # shared volume private ip file
+                if not etc_file.resolve() == ctdb_file:
+                    delete_it = True
+        else:
+            symlink_it = True
+
+        if delete_it:
+            try:
+                etc_file.unlink()
+            except Exception:
+                self.logger.error('Failed deleting %s', str(etc_file), exc_info=True)
+                return False
+
+        if symlink_it:
+            try:
+                etc_file.symlink_to(ctdb_file)
+            except Exception:
+                self.logger.error('Failed symlinking %s to %s', str(etc_file), str(ctdb_file), exc_info=True)
+                return False
+
+        # ctdb documentation is _VERY_ explicit about
+        # how this file should be modified
+        if not delete:
+            # in the case of adding a node, it _MUST_ be
+            # added to the end of the file always
+            with open(PRI_IP_FILE, 'a') as f:
+                f.write(pri_ip + '\n')
+                f.flush()
+        else:
+            with open(PRI_IP_FILE) as f:
+                lines = f.read().splitlines()
+
+            # in the case of removing a node, a "#"
+            # _MUST_ be put in front of the entry in
+            # the file and the other lines _MUST_ be
+            # kept in the same order when writing it
+            # back out or ctdb will barf and fail to
+            # reload...
+            index = lines.index(pri_ip)
+            lines[index] = '#' + pri_ip
+
+            with open(PRI_IP_FILE, 'w') as f:
+                f.writelines(map(lambda x: x + '\n', lines))
+                f.flush()
+
+        return True
+
+    @accepts(Dict(
+        'node_create',
+        IPAddr('ip'),
+    ))
+    @job(lock=PRI_LOCK)
+    async def create(self, job, data):
+        """
+        Add a ctdb private address to the cluster
+
+        `ip` is an IP v4/v6 address
+        """
+
+        schema_name = 'node_create'
+        verrors = ValidationErrors()
+
+        await self.middleware.call(
+            'ctdb.private.ips.common_validation', data['ip'], schema_name, verrors
+        )
+        if not await self.middleware.call('ctdb.private.ips.update_file', data['ip'], verrors):
+            raise CallError(f'Failed updating {PRI_IP_FILE}. Please check logs.')
+
+        return data['ip']
+
+    @accepts(Dict(
+        'node_delete',
+        IPAddr('ip'),
+    ))
+    @job(lock=PRI_LOCK)
+    async def delete(self, job, data):
+        """
+        Delete a Private IP address from the ctdb cluster.
+
+        `ip` is an IP v4/v6 address
+        """
+
+        schema_name = 'node_delete'
+        verrors = ValidationErrors()
+
+        await self.middleware.call(
+            'ctdb.private.ips.common_validation', data['ip'], schema_name, verrors, True
+        )
+        if not await self.middleware.call('ctdb.private.ips.update_file', data['ip'], verrors, True):
```

Comment:
Wouldn't it be better to make ctdb.private.ips.update_file raise an error show we could potentially know the error instead of telling user to check logs?